Make Monex compatible with eXist-db 6.x.x and 7.x.x

adamretter opened the pull request:
I worked out how to have a single codebase for Monex and make it work with eXist-db 6.x.x and 7.0.0-SNAPSHOT (and therefore by extension Elemental 6.x.x and 7.x.x).

All tests are passing against 6.x.x and 7.x.x.

**NOTE** - This first requires a version 2.0.0 of [exist-apps-parent](https://github.com/eXist-db/exist-apps-parent) to be produced by merging the PR - https://github.com/eXist-db/exist-apps-parent/pull/93 and creating a release of it.

**NOTE** PR https://github.com/eXist-db/monex/pull/303 should be merged to `master` first, and then this PR should be rebased atop `master` before being re-pushed and merged.

**NOTE** The `CI / Integration Test (DB API 7.x.x)` currently fails here as the latest SNAPSHOT of Elemental is not available to CI yet, but if you build and test locally you will see this does also pass.

Closes https://github.com/eXist-db/monex/pull/293
Closes https://github.com/eXist-db/monex/pull/285

adamretter committed:
[feature] Make Monex compatible with eXist-db and Elemental versions 6.x.x and 7.x.x
See https://github.com/evolvedbinary/elemental/issues/6

adamretter committed:
[bugfix] Correct the LGPL 2.1 license headers so they are correct, consistent, and follow eXist-db

adamretter commented on package.json:
```diff
@@ -1,7 +1,11 @@
+  "version": "4.3.0-SNAPSHOT",
   "description": "Monitoring Application for eXist-db",
+  "engines": {
+    "node": ">=18.12.1 <19",
```
@duncdrum this version was already specified in the pom.xml. I just copied it into the package.json from there. Do you want to send a separate PR to update the Node.js version?

adamretter replied:
@duncdrum I have now moved the commit for that out of this PR and into a separate PR here - https://github.com/eXist-db/monex/pull/302

duncdrum commented on pom.xml:
```diff
@@ -137,329 +238,187 @@
+                                <goals>
+                                    <goal>exec</goal>
+                                </goals>
+                                <configuration>
```
insert BaseUrl here

adamretter commented:
> . We actually decided to not maintain a single version compatible with v6 and v7

@duncdrum Will that be re-considered? IMHO Having two separate versions, branches and releases doesn't make a lot of sense when 6.x.x and 7.x.x are almost identical, and it will cost more you with a higher maintenance overhead.

> have a separate branch for work on the v7 release see https://github.com/eXist-db/monex/pull/285 .

It looks like the Monex v7 branch is very incomplete still and doesn't yet work with eXist-db 7.0.0-SNAPSHOT. IMHO It's better to take this one which is simpler and is ready to go *now*.

marmoure reviewed (approved):
LGTM, Thanks @adamretter for PR, the CI changes are much appreciated.
I have built and tested with locally with exist-db 6.1.0 and 6.4.0 and it working well.
I do hope we can get the CI to green as soon as possible, and keep it that way from now on.

reinhapa reviewed (changes requested):
In contrast to @duncdrum I could see this change as an intermittent step on the way...

adamretter commented on README.md:
````diff
@@ -81,31 +66,45 @@ In the Monex Remote Monitoring tab click "Run" to start all remote monitoring jo
+  mvn verify -P docker-integration-test-db-api-6,cypress-with-recording-db-api-6
+  ```
+
+  2. Against Elemental 7.x.x:
+  ```bash
+  mvn verify -P docker-integration-test-db-api-7,cypress-with-recording-db-api-7
+  ```
````
This can work with eXist-db 7.0.0-SNAPSHOT as well. However as there is no 7.0.0 release of eXist-db, we have used Elemental 7 here to show that this works. We did document this in a couple of places also, see: https://github.com/eXist-db/monex/pull/298/files#diff-9c5fb3d1b7e3b0f54bc5c4182965c4fe1f9023d449017cece3005d3f90e8e4d8R377

dizzzz commented:
@adamretter regarding compatibility, you made quite some changes in the JMX servlet. Are the proposed changes compatible with eXist?

adamretter commented on monex-parent/pom.xml:
```diff
@@ -0,0 +1,213 @@
+
+    <parent>
+        <groupId>xyz.elemental.expath</groupId>
+        <artifactId>elemental-expath-package-parent</artifactId>
```
Yes, the previous eXist-db one has too many problems. The details of that are discussed separately here - https://github.com/eXist-db/expath-crypto-module/pull/87#issuecomment-2999135156

dizzzz commented:
@adamretter if you may recall, you made changes in the JMX servlet recently.

dizzzz reviewed (changes requested):
build fails

adamretter commented:
> build fails

Yes, that is actually expected and a good thing. I wrote about that here - https://github.com/eXist-db/monex/pull/298

duncdrum reviewed (changes requested):
Ok I m not even half-way through the changes. 

- Tests are failing for v7, since that is the whole point of this PR, that's a blocker. 
- This PR could be very much simplified, if it wouldn't try to combine v6 and v7. Which we already decided not to do, even after reconsideration. I suggest focusing on the changes from `db-api-7` which are necessary for v7 and target the WIP branch I already mentioned.  If there is a need for `db-api-6` changes let's deal with them separately. 
- Any declared processor dependency must also be tested. Irrespective of processor. I m happy to discuss options here. But as it stands elemental v6 and exist v7 are declared but not tested. This is another blocker. 
- The introduction of multiple parent poms increases the maintenance burden, we should simplify not add complexity. 
- While we are patiently waiting for a response to the license issue around elemental
  - Demanding PR reviews on slack is not helping
-

adamretter commented on .gitignore:
```diff
@@ -19,76 +8,25 @@ pom.xml.versionsBackup
-
-##############################
-## Node / JS
-##############################
```
@duncdrum They are not included in the repo. In this PR instead there is a bugfix so that instead of writing them into the `src` folder, we corrected the config so they are written into the `target` folder (which is excluded in `.gitignore`)

adamretter committed:
[ci] Run integration tests as a Matrix Job to enable concurrency

adamretter commented on cypress.json:
```diff
@@ -1,12 +1,13 @@
 {
-  "baseUrl": "http://localhost:8080/exist",
+  "baseUrl": "http://localhost:10086/exist",
```
@duncdrum Three good reasons:

1. When running the integration tests, if you already have eXist-db running on your system elsewhere (as many of us do), then the tests fail as eXist-db in Docker for this project cannot be started on the same port

2. You cannot use the same port twice, if you want to run the Integration Tests in parallel, which is exactly what happens if you are using a modern concurrent enabled version of Maven, then the two integration tests can collide as they try to use the same port.

3. It is generally accepted good practice to use a different port during integration testing so it doesn't clash with any other instances. Normally this would be the next free randomly assigned port, but there isn't a good mechanism to use a random port number with the docker-maven-plugin. We have taking this approach for a long time in the main eXist-db project (for an example, see: https://github.com/eXist-db/exist/blob/develop/exist-core/src/main/java/org/exist/test/ExistWebServer.java#L126C74-L126C86)

duncdrum replied:
We test all stock apps against the default ports. I'm quite aware of how this works. Thanks for the three paragraph explanation. Random ports hidden inside maven plugins will make testing this much more cumbersome, and inconsistent with the other apps. If you have an exist running on 8080 simply install what you want to test there. The problem you are painting doesn't exist.

dizzzz commented:
Aside of the group-id/artifact-id in the parent (which has been addressed), I have the following remarks.

- it is not easy possible to see the difference between de v6 and v7 classes, because the version is encoded in the Filenames. It would be easier to maintain and review and ... when the classes have the same name but are put in different packages... org.exist.console.xquery.v6 and org.exist.console.xquery.v7
- exist-db does not use yet the latest jetty version, meaning that v7 can not be tested and validated yet; this is a risk.

adamretter committed:
[refactor] Switch to a newer eXist-db EXPath Package Parent
NOTE - This first requires a version 2.0.0 of [exist-apps-parent](https://github.com/eXist-db/exist-apps-parent) to be produced by merging the PR - [eXist-db/exist-apps-parent#93](https://github.com/eXist-db/exist-apps-parent/pull/93) and creating a release of it.

dizzzz commented:
@adamretter 
1) thnx!
2) I understand the reasoning and potential confusion. Need to think. The issue I faced were (1) the difficulty to see the difference compared to the original code and (2) an easy check on the differences between 6 vs 7.
3) my wrong, I thought we still used the javax.servlet stuff, my now I see the 'import jakarta.servlet.http.HttpServlet;'

adamretter committed:
[bugfix] Correct the version of javax.websocket-api so that it matches eXist-db 6.4.0

adamretter committed:
[ci] Add further targets for testing in CI

duncdrum commented on monex-parent/pom.xml:
```diff
@@ -0,0 +1,213 @@
+
+    <properties>
+        <!-- Version of eXist-db and Elemental 6.x.x that Monex is compatible with -->
+        <db.api.6.version>6.0.0</db.api.6.version>
```
@adamretter my point exactly, a source-code collection of 1704 items for users to dig through counts as undocumented. Especially when `api version`, `Java API`,  nor `6.0.0` appear in that collection of items in a relevant context.

adamretter replied:
@duncdrum I am sorry, but I cannot understand what you are saying.

How is the API of eXist-db undocumented exactly? There is Javadoc on all public methods, and that Javadoc is also published every time a release of eXist-db is made.

What is the concrete problem with documentation that you see here (that did not already exist before this PR)?

adamretter commented:
@duncdrum That's because eXist 7.0.0-SNAPSHOT needs some bug fixes to make it work again with Monex. We already made those in Elemental 7.2.0, you can take those if you like.

adamretter commented:
>  I m not sure how I would take bug fixes to make your pr compatible with exist 7

@duncdrum It's linked in the updates from GitHub above! You simply cherry-pick the commits from this PR and include our LGPL 2.1 license header in each file you change: https://github.com/evolvedbinary/elemental/pull/10

> Bug fixes are of course always welcome.

As explained before, due to reasons outside of our control, we (Evolved Binary) will no longer contribute code directly to eXist-db, only to the Apps and Supporting Libraries.

duncdrum reviewed (changes requested):
There are still problems: 
- it doesn't run on `:latest` aka exist7. As long as it doesn't there is nothing to review.
- random ports break convention with all other apps. Testing and will make testing harder
- Why was our copyright changed to 2017?
- Node version should be `22`
- I still consider the combined v6 v7 build overly complex and a maintenance nightmare.

adamretter commented:
> This PR seems to be blocked now. I would have at least expected an actionable issue description on exist-db.

@line-o I am not clear what you mean... can you explain please?

dizzzz reviewed (changes requested):
PR is *not* compatible with eXist-db 7 - SNAPSHOT

line-o commented:
When testing this branch locally I found the hard-coded port number in the base path to be very inconvenient. Especially since it was _not_ the standard port number 8080.

I propose that the cypress tests read the port from an environment variable that can be set per run but they default to 8080.

adamretter commented:
> PR is not compatible with eXist-db 7 - SNAPSHOT

@dizzzz Yes! As already stated [here](https://github.com/eXist-db/monex/pull/298#issuecomment-3130763528) - there is a bug in eXist-db 7.0.0-SNAPSHOT that stops it working correctly with Monex (as it is today in the last release of Monex). This was previously fixed in Elemental. Someone in the eXist-db team needs to fix that in eXist-db

dizzzz commented on db-api/db-api-7/pom.xml:
```diff
@@ -0,0 +1,127 @@
+        </dependency>
+        <dependency>
+            <!-- NOTE(AR) (Inherited from monex-parent) Could be swapped out for eXist-db 7.0.0 if it is released, but leaving it as Elemental does no harm either -->
+            <groupId>xyz.elemental.fork.org.exist-db</groupId>
```
My report is actually complete. It is all about a Type conflict.

line-o commented:
The community decided to take a different route as exist-db 7 will expose an API that monex is just consuming from here on.

@adamretter Thank you for making the effort to create a version of monex that would be compatible with exist-db 6 and elemental. Since the PR in its current state was never compatible with any SNAPSHOT of exist-db 7 we decided to take this different route. We believe it will be easier to maintain monex' functionality going forward by consuming a well defined API and being a mere frontend for core functionality.

adamretter commented:
>  Since the PR in its current state was never compatible with any SNAPSHOT of exist-db 7

Sorry but that statement has never been true. I dispelled that at many meetings we had, and demo'd it working too at one of the meetings. It's sad that that myth continues to be perpetrated.

> The community decided to take a different route as exist-db 7 

That is not a representative statement of the community as several people including the creator of Monex wished to move forward with this. I think you instead mean that 2 of the more recent Monex developers decided to take a different route. Which is fine, I wish you and them luck.
